### PR TITLE
HeaderCheck: Let's make sure that our headers actually compile!

### PR DESCRIPTION
**Pull request by BenWiederhake**
Let's say we have a file `compile_me.cpp`, like this:

```c++
// file: compile_me.cpp
#include <LibDNS/Question.h>
// That's all!
```

Before this PR, it didn't compile.

HeaderCheck is a tool that automatically tries to compile all headers this way. (All headers in `Userland/Libraries/`.) However, doing so takes a long time, and it sucks when a CI build fails due to this after a long time. [1](https://discord.com/channels/830522505605283862/830525031720943627/943589070565679155) [2](https://github.com/SerenityOS/serenity/commit/bc98ad9cc1452a3db574eca44b029c2dfca43605) [3](https://discord.com/channels/830522505605283862/830739873119207426/1019298311590645792).

This PR:
- fixes all of the existing problems, including the nasty dependency cycle in LibWeb involving `CSS::Length`
- includes AK into the HeaderCheck scope
- runs HeaderCheck as a weekly cronjob that runs at 2am every morning

Here's a preview how this will look: https://github.com/BenWiederhake/serenity/actions/runs/3048260526/jobs/4913127822

Notice that it only takes 8 minutes (plus 14 minutes to build the toolchain from scratch), so I must once again recommend that HeaderCheck be run on every CI execution, but oh well.

And just to hammer down the point that this problem isn't imaginary: people sometimes have to first fiddle with unrelated headers in order to do their work: https://github.com/SerenityOS/serenity/pull/15228/commits/6089d4e56a5827a12098b168eb9822d26f524ec7

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/StringPrototype.h [hunk 1]
```diff
@@ -17,7 +17,7 @@ struct CodePoint {
 };
 
 CodePoint code_point_at(Utf16View const& string, size_t position);
-static constexpr Utf8View whitespace_characters = Utf8View("\x09\x0A\x0B\x0C\x0D\x20\xC2\xA0\xE1\x9A\x80\xE2\x80\x80\xE2\x80\x81\xE2\x80\x82\xE2\x80\x83\xE2\x80\x84\xE2\x80\x85\xE2\x80\x86\xE2\x80\x87\xE2\x80\x88\xE2\x80\x89\xE2\x80\x8A\xE2\x80\xAF\xE2\x81\x9F\xE3\x80\x80\xE2\x80\xA8\xE2\x80\xA9\xEF\xBB\xBF"sv);
```
Comment (davidot) [inline Userland/Libraries/LibJS/Runtime/StringPrototype.h:20]: Isn't `explicit constexpr Utf8View(StringView string)` a valid constructor for this?

In fact I think that static constexpr forces it to be constexpr but that I'm not sure about.

Comment (BenWiederhake) [reply Userland/Libraries/LibJS/Runtime/StringPrototype.h:20]: You're right! Turns out, I misread the compilation error message, and it's missing an include instead.

Weird, I could have sworn that the compilation error message said something else.

**Comment by BenWiederhake** [comment]
Changes since last push:
- Resolved the recent conflict in `LibWeb/CSS/GridTrackSize.{cpp,h}`
- Dropped the name change of `whitespace_characters`, as suggested by @davidot 
- Set to draft: The first commit alone does some pretty funky changes. Let's review that stand-alone first, in https://github.com/SerenityOS/serenity/pull/15236 . Once that's been dealt with, I'll un-draft this PR.

File: Meta/HeaderCheck/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,5 @@
 execute_process(COMMAND "${CMAKE_CURRENT_SOURCE_DIR}/generate_all.py" "${SERENITY_ARCH}" OUTPUT_VARIABLE SOURCES_STRING)
 string(REPLACE "\n" ";" SOURCES_LIST ${SOURCES_STRING})
 
+add_compile_options(-Wno-psabi)
```
Comment (trflynn89) [inline Meta/HeaderCheck/CMakeLists.txt:4]: It'd be nice for the commit to explain why this is needed

File: Meta/HeaderCheck/generate_all.py [hunk 3]
```diff
@@ -12,7 +12,9 @@
 
 
 def get_headers_here():
-    result = subprocess.run(['git', 'ls-files', 'Userland/Libraries/*.h'], check=True, capture_output=True, text=True)
+    result = subprocess.run(
```
Comment (trflynn89) [inline Meta/HeaderCheck/generate_all.py:15]: This seems like it should be in its own commit

File: Userland/Libraries/LibWeb/WebGL/WebGLRenderingContextBase.h [hunk 4]
```diff
@@ -10,7 +10,6 @@
 #include <AK/WeakPtr.h>
 #include <AK/Weakable.h>
 #include <LibGL/GLContext.h>
-#include <LibJS/Heap/GCPtr.h>
```
[the remark this answers is not in the document]
Comment (BenWiederhake) [reply Userland/Libraries/LibWeb/WebGL/WebGLRenderingContextBase.h:13]: It's already included anyway through `PlatformObject.h`, but yeah, I see the benefit of including it directly for clarity. I'll add it back.

File: Meta/HeaderCheck/CMakeLists.txt (hunk 2 above)
Comment (BenWiederhake) [reply Meta/HeaderCheck/CMakeLists.txt:4]: LibSoftGPU does not compile without it, as evidenced by `Userland/Libraries/LibSoftGPU/CMakeLists.txt:9`.

File: Userland/Libraries/LibWeb/WebGL/WebGLRenderingContextBase.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (trflynn89) [reply Userland/Libraries/LibWeb/WebGL/WebGLRenderingContextBase.h:13]: :+1: yes, let's not rely on transitive includes

File: Meta/HeaderCheck/CMakeLists.txt (hunk 2 above)
Comment (trflynn89) [reply Meta/HeaderCheck/CMakeLists.txt:4]: Right, just saying that should be in the commit message

File: Meta/HeaderCheck/generate_all.py (hunk 3 above)
Comment (BenWiederhake) [reply Meta/HeaderCheck/generate_all.py:15]: That feels like excessive splitting, but okay, will do :)

**Comment by BenWiederhake** [comment]
Changes since last push:
- Rebased for convenience
- Split the AK header move and HeaderCheck scop extension into separate commits, as suggested
- Added back in an indirect include, as suggested
- Added an explanation why HeaderCheck needs `-Wno-psabi` (in short: because LibSoftGPU currently requires it), as suggested

File: .github/workflows/headercheck.yml [hunk 5]
```diff
@@ -0,0 +1,73 @@
+name: Check that headers compile (Libraries and Services)
+
+on:
+  # Automatically run at the end of every day.
```
Comment (AtkinsSJ) [inline .github/workflows/headercheck.yml:4]: Comment is wrong. :^)

File: .github/workflows/headercheck.yml [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+      - name: "Install Ubuntu dependencies"
+        # These packages are already part of the ubuntu-22.04 image:
+        # cmake libgmp-dev npm shellcheck
+        # Packages below aren't.
+        run: |
+          wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+          sudo add-apt-repository 'deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-14 main'
+          sudo apt-get update
+          sudo apt-get install -y clang-format-14 ccache gcc-11 g++-11 libstdc++-11-dev libmpfr-dev libmpc-dev ninja-build unzip
```
Comment (AtkinsSJ) [inline .github/workflows/headercheck.yml:36]: `clang-format-14` feels like it shouldn't be necessary? Though I don't know if it makes much difference.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Looks good. The action file I'm less sure about, but mostly because of my own ignorance.

File: .github/workflows/headercheck.yml (hunk 5 above)
Comment (BenWiederhake) [reply .github/workflows/headercheck.yml:4]: Doink! You're right, I'll rewrite that.

File: .github/workflows/headercheck.yml (hunk 6 above)
Comment (BenWiederhake) [reply .github/workflows/headercheck.yml:36]: You're right! I'll also remove ccache, since obviously this build process shouldn't build the same compilation unit twice. I'll keep unzip, because I'm not entirely sure where it's used, and I guess it's a rather unproblematic dependency. The others are most definitely used.

File: Meta/HeaderCheck/CMakeLists.txt (hunk 2 above)
Comment (BenWiederhake) [reply Meta/HeaderCheck/CMakeLists.txt:10]: Both commit message and a comment in CMakeLists.txt now explain what's going on.

**Comment by bgianfo** [comment]
Can this go in without the weekly pipeline? I fear this will just rot once you stop looking at it. It's a good goal, but I don't think automating this incantation will buy us mutch besides some noise that most folks will just ignore?

Thoughts?

**Comment by BenWiederhake** [comment]
Sigh, how frustrating. How come I seem to be the only one interested in making the library headers compilable? I mean, I know part of the answer, but I find it surprising nevertheless.

I guess then I'll remove it. :(

**Comment by BenWiederhake** [comment]
Changes since last push:
- Rebased (especially interesting because #15235 landed)
- Dropped the commit that establishes a weekly workflow; see previous discussion :(

**Review by trflynn89 — COMMENTED** [review]
Last comment, LGTM otherwise

File: AK/FuzzyMatch.h [hunk 7]
```diff
@@ -19,126 +18,12 @@ struct FuzzyMatchResult {
 static constexpr int const RECURSION_LIMIT = 10;
 static constexpr int const MAX_MATCHES = 256;
```
[the remark this answers is not in the document]
Comment (BenWiederhake) [reply AK/FuzzyMatch.h:19]: Done!

**Comment by BenWiederhake** [comment]
Changes since last push:
- Moved the two constexprs as suggested by trflynn89 (thanks!)
- Rebased for convenience
- Another rebuild just to make extra sure there are no "newly-missing" headers
- Independent rebuild with check-headers-acyclic.py (PR coming soon™) to make sure I didn't accidentally create a cycle

File: AK/FuzzyMatch.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (BenWiederhake) [reply AK/FuzzyMatch.h]: In case you're curious about the backstory: While making that edit, I initially left in the "short" version of the function, i.e. `fuzzy_match(String const& needle, String const& haystack)`. This function is just a thin wrapper around the full interface, using `MAX_MATCHES` in its definition. For some weird reason I also assumed that `RECURSION_LIMIT` would be the default recursion limit and somehow used in that function definition. At that point in time, it was a reasonably good idea to keep it in.

Then I decided to just move it out of line, because fuzzy_match has no business being so performance critical to slow down compilation that much, and I didn't notice that the constexpr-s became totally superfluous.

In case you're _not_ curious about the backstory, here are some happy cows: https://www.redditp.com/r/happycowgifs
